ml_genn/ml_genn/compilers/compiler.py

```diff
@@ -80,28 +82,71 @@ def create_reset_custom_update(reset_vars, var_ref_creator):
 
     # Loop through reset vars
     for name, type, value in reset_vars:
-        # Add variable reference using function to create variable reference
+        # Add variable reference
         model.add_var_ref(name, type, var_ref_creator(name))
 
-        # Add reset value parameter
-        model.add_param(name + "Reset", type, value)
+        # If variable should be reset to another variable
+        if isinstance(value, str):
+            # If variable to reset to isn't already in list
+            # **TODO** give warning if it's not afterwards!
+            existing_reset_var = [v for v in reset_vars if v[0] == value]
+            if len(existing_reset_var) == 0:
+                # Add read-only variable reference to other variable
+                model.add_var_ref(value, type, var_ref_creator(value))
+                model.set_var_ref_access_mode(value, VarAccessMode_READ_ONLY)
+
+            # Add code to set var
+            model.append_update_code(f"$({name}) = $({value});")
+        # Otherwise
+        else:
+            # Add reset value parameter
+            model.add_param(name + "Reset", type, value)
+
+            # Add code to set var
+            model.append_update_code(f"$({name}) = $({name}Reset);")
 
-        # Add code to set var
-        model.append_update_code(f"$({name}) = $({name}Reset);")
     return model
 
+class SupportedMatrixType:
+    def __init__(self, supported: List[int]):
+        # Build dictionary of supported connectivity types and order
+        self._supported = {v: i for i, v in enumerate(supported)}
+    
+    def get_best(self, available: List[int]) -> Optional[int]:
+        # Intersect supported connectivity types with those that are available
+        possible = self._supported.keys() & available
+        
+        # If there are no possible options
+        if len(possible) == 0:
+            return None
+        # Otherwise, return the connectivity with 
+        # the highest priority from possible
+        else:
+            return min(possible, key=lambda p: self._supported[p])
+        
+class ZeroInSyn(Callback):
+    def __init__(self, genn_syn_pop, example_timesteps: int):
+        self.genn_syn_pop = genn_syn_pop
+        self.example_timesteps = example_timesteps
+
+    def on_timestep_begin(self, timestep: int):
```

Comment:
I am obviously naive, but does this mean in_syn is reset at the beginning of the last timestep of a trial? Not at the end?

Reply:
That was my attempt to duplicate https://github.com/tnowotny/genn_eventprop/blob/main/simulator_SHD.py#L1238-L1244

Reply:
oh, yes, I had completely forgotten about that. Would it be worth copying the comment about clamping on last time step and  not bleeding spikes into the next trial?

Reply:
Yes, I had also clearly forgotten exactly what it was doing when I moved it out of EventProp - it's quite a specific thing so will move it back, rename and add a comment

Reply:
You clearly know the right loose threads to tug on! The other use of ``ZeroInSyn`` in ``InferenceCompiler`` was also flawed - now fixed